File: CPlusPlus/Sample_Source/Text/ExtractTextByPatternMatch/ExtractTextByPatternMatch.cpp
//
// Copyright (c) 2023, Datalogics, Inc. All rights reserved.
//
// For complete copyright information, refer to:
// http://dev.datalogics.com/adobe-pdf-library/license-for-downloaded-pdf-samples/
//
// This sample extracts text that matches a given pattern using regular expressions (regex)
// in a PDF document and saves the text to a file.
//
// For more detail see the description of the ExtractTextByPatternMatch sample program on our Developer’s site,
// http://dev.datalogics.com/adobe-pdf-library/sample-program-descriptions/c1samples

#include <fstream>
#include <string>

#include "InitializeLibrary.h"
#include "APDFLDoc.h"
#include "DLExtrasCalls.h"
#include "TextExtract.h"

const char *DEF_INPUT = "../../../../Resources/Sample_Input/ExtractTextByPatternMatch.pdf";
const char *DEF_OUTPUT = "ExtractTextByPatternMatch-out.txt";

// This sample will look for text that matches a phone number pattern
const char *DEF_PATTERN = regexPattern[PHONE_PATTERN];

int main(int argc, char **argv) {

    // Initialize the library
    APDFLib libInit;
    ASErrorCode errCode = 0;
    if (libInit.isValid() == false) {
        errCode = libInit.getInitError();
        std::cout << "Initialization failed with code " << errCode << std::endl;
        return libInit.getInitError();
    }

    DURING

        APDFLDoc inAPDoc(DEF_INPUT, true);
        std::ofstream outputFile(DEF_OUTPUT);

        // Use PDWordfinder to find all the Words in our input document
        PDWordFinderConfigRec wfConfig;
        memset(&wfConfig, 0, sizeof(PDWordFinderConfigRec));
        wfConfig.recSize = sizeof(PDWordFinderConfigRec);

        // Need to set this to true so phrases will be concatenated properly
        wfConfig.noHyphenDetection = true;

        // Get the matches for this document
        PDDocTextFinder matchFinder = PDDocTextFinderCreate(&wfConfig);
        PDDocTextFinderMatchList matchList =
            PDDocTextFinderAcquireMatchList(matchFinder, inAPDoc.getPDDoc(), PDAllPages, 0, DEF_PATTERN);

        // Iterate over the matches that were found by PDDocTextFinder and print them out in the
        // output file.
        for (size_t matchInstance = 0; matchInstance < matchList.numMatches; ++matchInstance) {

            PDDocTextFinderMatchRec match = matchList.matches[matchInstance];

            // Put this matched phrase in the output document
            outputFile << match.phrase << std::endl;
        }

        // Release resources here.
        PDDocTextFinderReleaseMatchList(matchFinder);
        PDDocTextFinderDestroy(matchFinder);

        // Close the output file.
        outputFile.close();

    HANDLER
        errCode = ERRORCODE;
        libInit.displayError(errCode);
    END_HANDLER

    return errCode;
}

File: DotNET/Sample_Source/Text/ExtractTextByPatternMatch/ExtractTextByPatternMatch.cs
using System;
using System.Collections.Generic;
using Datalogics.PDFL;

/*
 * This sample extracts text that matches a given pattern in a PDF
 * document and saves the text to a file.
 *
 * Copyright (c) 2022-2023, Datalogics, Inc. All rights reserved.
 *
 * For complete copyright information, refer to:
 * http://dev.datalogics.com/adobe-pdf-library/license-for-downloaded-pdf-samples/
 *
 */
namespace ExtractTextByPatternMatch
{
    class ExtractTextByPatternMatch
    {
        static void Main(string[] args)
        {
            Console.WriteLine("ExtractTextByPatternMatch Sample:");


            using (new Library())
            {
                Console.WriteLine("Initialized the library.");

                // Set Defaults
                String sInput = Library.ResourceDirectory + "Sample_Input/ExtractTextByPatternMatch.pdf";
                String sOutput = "../ExtractTextByPatternMatch-out.txt";
                String sPattern = "((1-)?(\\()?\\d{3}(\\))?(\\s)?(-)?\\d{3}-\\d{4})";  // phone numbers

                using (Document doc = new Document(sInput))
                {
                    int nPages = doc.NumPages;

                    Console.WriteLine("Input file:  " + sInput);

                    using (WordFinderConfig wordConfig = new WordFinderConfig())
                    {
                        // DocTextFinder creates a string of characters to search.
                        // This string may contain hyphens and the regular expression
                        // may contain hyphens.  So we do not want the wordfinder to
                        // alter that or else the search may fail.
                        wordConfig.NoHyphenDetection = true;

                        using (System.IO.StreamWriter outfile = new System.IO.StreamWriter(sOutput))
                        {
                            Console.WriteLine("Writing to: " + sOutput);

                            // Create a DocTextFinder with the wordfinder.
                            using (DocTextFinder docTextFinder =
                                new DocTextFinder(doc, wordConfig))
                            {
                                // Retrieve the phrases matching the pattern.
                                IList<DocTextFinderMatch> docMatches =
                                    docTextFinder.GetMatchList(0, nPages - 1, sPattern);

                                // Iterate through the matches and write them out to the file.
                                foreach (DocTextFinderMatch match in docMatches)
                                {
                                    outfile.WriteLine(match.MatchString);
                                }
                            }
                        }
                    }
                }
            }
        }
    }
}

